src/transformers/modeling_utils.py

```diff
@@ -812,6 +812,62 @@ def enforce_repetition_penalty_(self, lprobs, batch_size, num_beams, prev_output
                 else:
                     lprobs[i, previous_token] /= repetition_penalty
 
+    def finalize_generation_logscores(
```

Comment:
I like this!

Reply:
I would call this 
`postprocess_generation_logscores` cause we have some notion of finalized hypotheses.
But I love this refactor. The `generate` fn was/is way too long to follow and this helps a significant amount.

Reply:
@yjernite - Actually taking a closer look it this, I'm not so sure if we even should follow suite with a unified sampling function. As it is implemented now, `_generate_no_beam_search` applies two softmax operations (log-softmax and softmax) instead of only softmax (as it was before) so this is a big breaking change for all models doing sampling (GPT2 generation). In beam search, sampling is very rarely used, but in `no_beam_search` it is heavily used and I'm not sure we want a breaking change here. 

The tests don't capture that because we don't have a sampling test for GPT2 only a greedy test (we should prob add a sampling test for GPT2).

Reply:
Okey that took me some time to realize :D So because `softmax(log_softmax(X))` is equal to `softmax(X)`, there is not even a breaking change here. But now we calculate the softmax twice instead of only once as before, is the extra cost that insignificant @yjernite ?